File: 89.格雷编码.py
#
# @lc app=leetcode.cn id=89 lang=python3
#
# [89] 格雷编码
#

# @lc code=start
class Solution:
    def grayCode(self, n: int) -> List[int]:
        if n==0:
            return [0]
        if n==1:
            return [0,1]
        result=self.grayCode(n-1)
        for i in range(len(result)-1,-1,-1):
            result.append(result[i]+2**(n-1))
        return result
# @lc code=end

File: 89.格雷编码.cpp
/*
 * @lc app=leetcode.cn id=89 lang=cpp
 *
 * [89] 格雷编码
 */

// @lc code=start
class Solution {
public:
    vector<int> grayCode(int n) {
        if(n==0)
            return vector<int>(1,0);
        if(n==1){
            vector<int> result(pow(2,n),0);
            result[0]=0;
            result[1]=1;
            return result;
        }
        vector<int> result=grayCode(n-1);
        for(int i=result.size()-1;i>=0;i--){
            result.push_back(result[i]+pow(2,n-1));
        }
        return result;
    }
};
// @lc code=end

